### PR TITLE
New Next Permutation Goodies

### PR DESCRIPTION
**Pull request by Hasil-Sharma**
Instant Answer Page: https://duck.co/ia/view/generating_next_permutation

### CODE REVIEW

**Comment by daxtheduck** [comment]
@Hasil-Sharma Hey!
        
Thanks for taking the time to contribute! We really appreciate it.

We work closely with every contributor to make Instant Answers the best they can be, so we appreciate your patience as we look over your code. From here, the process usually goes like this:

1. Pull Request is reviewed by the DuckDuckGo staff and community
2. Staff and community will leave feedback with any necessary updates to the function or design.
3. Once you've made any necessary corrections, then your Instant Answer will be merged and deployed live on DuckDuckGo!

If you have any questions along the way, feel free to ask them here. Our staff and community are also available on Slack to answer any questions you may have. If you'd like to join us there please head to https://quackslack.herokuapp.com/ to get an invite.

>More Info: http://docs.duckduckhack.com/submitting/submitting-overview.html

Thanks!

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by Hasil-Sharma** [comment]
@GuiltyDolphin : I've created this PR ... Please review

File: lib/DDG/Goodie/NextPermutation.pm [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+# Start at http://docs.duckduckhack.com/walkthroughs/calculation.html if
+# you are new to instant answer development
+
+use DDG::Goodie;
+use strict;
+use warnings;
+
+zci answer_type => 'next_permutation';
+
+# Caching - http://docs.duckduckhack.com/backend-reference/api-reference.html#caching`
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/NextPermutation.pm]: Lets remove the auto-generated comments.

File: lib/DDG/Goodie/NextPermutation.pm [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    
+    #return unless (regex for sequence of only numbers)
+    
+    #Converting String to Array
+    my @array = split('',$_);
+    
+    #Finding the index and index + 1 to be swapped
+    my $index = swapable_index(@array);
+    
+    #Return unless there one such index
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/NextPermutation.pm]: We don't need all these comments.

File: lib/DDG/Goodie/NextPermutation.pm [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+# Triggers - http://docs.duckduckhack.com/walkthroughs/calculation.html#triggers
+triggers startend => 'next permutation', 'next perm';
+
+# Handle statement
+handle remainder => sub {
+    
+    #return unless (regex for sequence of only numbers)
+    
+    #Converting String to Array
+    my @array = split('',$_);
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/NextPermutation.pm]: We should be using something like `List::Permute` to get the permutations.

File: t/NextPermutation.t [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    'next permutation abcd' => build_test('abcd','abdc'),
+    'next permutation dhck' => build_test('dhck', 'dhkc'),
+    'next permutation dkhc' => build_test('dkhc', 'hcdk'),
+    'next permutation ab' => build_test('ab', 'ba'),
+    'next permutation 218765' => build_test('218765', '251678'),
+    # Try to include some examples of queries on which it might
+    # appear that your answer will trigger, but does not.
+    'next permutation 54321' => undef, 
+    'next permutation bb' => undef,
+    'next permutation 4321' => undef,
```
Comment (GuiltyDolphin) [inline t/NextPermutation.t]: Add a test for `'next permutation'` (non-trigger).

File: lib/DDG/Goodie/NextPermutation.pm (hunk 3 above)
Comment (Hasil-Sharma) [inline lib/DDG/Goodie/NextPermutation.pm]: I did look into one of the [perl module](https://metacpan.org/pod/Algorithm::Permute) that came up during our initial discussion, input given to that module returns all the possible permutations of the given sequence. For instance, input for sequence `1243` returns output in following order, but the correct output should be `1324` which comes much later in the output order

> 3, 4, 2, 1
4, 3, 2, 1
4, 2, 3, 1
4, 2, 1, 3
3, 2, 4, 1
2, 3, 4, 1
2, 4, 3, 1
2, 4, 1, 3
3, 2, 1, 4
2, 3, 1, 4
2, 1, 3, 4
2, 1, 4, 3
3, 4, 1, 2
4, 3, 1, 2
4, 1, 3, 2
4, 1, 2, 3
3, 1, 4, 2
1, 3, 4, 2
1, 4, 3, 2
1, 4, 2, 3
3, 1, 2, 4
1, 3, 2, 4
1, 2, 3, 4
1, 2, 4, 3

File: lib/DDG/Goodie/NextPermutation.pm (hunk 1 above)
Comment (Hasil-Sharma) [inline lib/DDG/Goodie/NextPermutation.pm]: Done

File: lib/DDG/Goodie/NextPermutation.pm (hunk 2 above)
Comment (Hasil-Sharma) [inline lib/DDG/Goodie/NextPermutation.pm]: Removed

File: t/NextPermutation.t (hunk 4 above)
Comment (Hasil-Sharma) [inline t/NextPermutation.t]: Done

**Comment by Hasil-Sharma** [comment]
@GuiltyDolphin

File: lib/DDG/Goodie/NextPermutation.pm [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    
+    @array = sort_index(\@array, $index);
+
+    my $result = join('', @array);
+
+    return "Next Permutation of the Sequence is $result",
+        structured_answer => {
+            data => {
+                title    => $result,
+                subtitle => 'Lexicographically Next Permutation of Sequence'
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/NextPermutation.pm]: We should include the sequence in the subtitle.

File: lib/DDG/Goodie/NextPermutation.pm [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    return @array;
+}
+
+sub min_greatest{
+    my @array = @{$_[0]};
+    my $start_index = $_[1];
+    my $end_index = $#array;
+
+    while($end_index > $start_index){
+        if($array[$end_index] gt $array[$start_index]){ last; }
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/NextPermutation.pm]: `last if $array[$end_index] gt $array[$start_index];`

File: lib/DDG/Goodie/NextPermutation.pm [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        $end_index = $end_index - 1;
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/NextPermutation.pm]: `$end_index--`

File: lib/DDG/Goodie/NextPermutation.pm [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+zci answer_type => 'next_permutation';
+
+zci is_cached => 1;
+
+triggers startend => 'next permutation', 'next perm', 'next permuta', 'next permu';
+
+handle remainder => sub {
+   
+    return unless /(^\s*[A-za-z]+\s*$|^\s*[0-9]+\s*$)/;
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/NextPermutation.pm]: `^([A-z]+|[0-9]+)$` (whitespace should be normalized already).

File: lib/DDG/Goodie/NextPermutation.pm [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+use DDG::Goodie;
+use strict;
+use warnings;
+
+zci answer_type => 'next_permutation';
+
+zci is_cached => 1;
+
+triggers startend => 'next permutation', 'next perm', 'next permuta', 'next permu';
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/NextPermutation.pm]: Why `next permuta` and `next permu`?

Comment (Hasil-Sharma) [inline lib/DDG/Goodie/NextPermutation.pm]: I was trying to account for typos, should I remove them ?

File: lib/DDG/Goodie/NextPermutation.pm (hunk 8 above)
Comment (Hasil-Sharma) [inline lib/DDG/Goodie/NextPermutation.pm]: Yes done

**Comment by Hasil-Sharma** [comment]
@GuiltyDolphin : I've done all the changes as suggested by you, please tell me how to proceed further 😄

File: lib/DDG/Goodie/NextPermutation.pm (hunk 9 above)
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/NextPermutation.pm]: Yes - we don't add triggers for typos - `perm` is fine as it is a common shortening.

Comment (Hasil-Sharma) [inline lib/DDG/Goodie/NextPermutation.pm]: Done

File: t/NextPermutation.t [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+            templates => {
+                group => 'text'
+            }
+        };
+}
+
+sub build_test { test_zci(build_structured_answer(@_)) }
+
+ddg_goodie_test(
```
Comment (sahildua2305) [inline t/NextPermutation.t]: @Hasil-Sharma you should add a couple of tests for `next perm` trigger as well 🙂

File: lib/DDG/Goodie/NextPermutation.pm [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+                subtitle => "Lexicographically Next Permutation For Sequence: $org_seq"
+            },
+            templates => {
+                group => 'text'
+            }
+        };
+};
+
+sub swapable_index{
+   my @array = @_;
```
Comment (sahildua2305) [inline lib/DDG/Goodie/NextPermutation.pm]: nit - there seems to be an issue with indentation here.

File: lib/DDG/Goodie/NextPermutation.pm [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+    my $end_index = $#array;
+
+    while($end_index > $start_index){
+        last if $array[$end_index] gt $array[$start_index];
+        $end_index--; 
+    }
+    
+    return $end_index;
+
+
```
Comment (sahildua2305) [inline lib/DDG/Goodie/NextPermutation.pm]: nit - extra blank lines

File: lib/DDG/Goodie/NextPermutation.pm [hunk 13]
```diff
@@ -0,0 +1,78 @@
+package DDG::Goodie::NextPermutation;
+
+
```
Comment (sahildua2305) [inline lib/DDG/Goodie/NextPermutation.pm]: nit - extra newlines

File: t/NextPermutation.t (hunk 10 above)
Comment (Hasil-Sharma) [inline t/NextPermutation.t]: @sahildua2305 : Added

File: lib/DDG/Goodie/NextPermutation.pm (hunk 13 above)
Comment (Hasil-Sharma) [inline lib/DDG/Goodie/NextPermutation.pm]: Removed

File: lib/DDG/Goodie/NextPermutation.pm (hunk 11 above)
Comment (Hasil-Sharma) [inline lib/DDG/Goodie/NextPermutation.pm]: removed

File: lib/DDG/Goodie/NextPermutation.pm (hunk 12 above)
Comment (Hasil-Sharma) [inline lib/DDG/Goodie/NextPermutation.pm]: removed

**Comment by Hasil-Sharma** [comment]
@sahildua2305  : Thanks for reviewing my PR, please help me in proceeding further :)

File: lib/DDG/Goodie/NextPermutation.pm [hunk 14]
```diff
@@ -0,0 +1,61 @@
+package DDG::Goodie::NextPermutation;
+#ABSTRACT : For a given sequence of numbers or alphabets (delimited by comma or space), generating the next permutation which will be lexicographically smallest
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/NextPermutation.pm]: This abstract should be shorter.

**Comment by Hasil-Sharma** [comment]
@GuiltyDolphin : I've reduced the abstract length

File: lib/DDG/Goodie/NextPermutation.pm [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+                title    => $result,
+                subtitle => "Lexicographically Next Permutation For Sequence: $org_seq"
+            },
+            templates => {
+                group => 'text'
+            }
+        };
+};
+
+sub swapable_index{
```
Comment (mintsoft) [inline lib/DDG/Goodie/NextPermutation.pm]: I think we could name this something better like `first_out_of_sort_order_index` perhaps

File: lib/DDG/Goodie/NextPermutation.pm [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+sub sort_index{
+    my @array= @{$_[0]};
+    my $index = $_[1];
+    my @subarray = @array[$index+1..$#array];
+    @subarray = sort { $a cmp $b} @subarray;    
+    splice @array, $index + 1, $#subarray + 1, @subarray;
+    return @array;
+}
+
+sub min_greatest{
```
Comment (mintsoft) [inline lib/DDG/Goodie/NextPermutation.pm]: I don't know what a `min_greatest` is, feels like a contradiction.

File: lib/DDG/Goodie/NextPermutation.pm [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+
+zci answer_type => 'next_permutation';
+zci is_cached => 1;
+triggers startend => 'next permutation', 'next perm';
+
+handle remainder => sub {
+    return unless /^([A-za-z]+|[0-9]+)$/;
+    my @array = split('',$_);
+    my $org_seq = $_;
+    my $index = swapable_index(@array);
```
Comment (mintsoft) [inline lib/DDG/Goodie/NextPermutation.pm]: I think all the naming problems stem from here, what does the index represent here?

**Comment by mintsoft** [comment]
I'll close this for now, if anything changes feel free to re-open

**Pull request closed without merging**